frame/pablo/src/lib.rs

```diff
@@ -467,6 +499,139 @@ pub mod pallet {
 		pub(crate) fn account_id(pool_id: &T::PoolId) -> T::AccountId {
 			T::PalletId::get().into_sub_account(pool_id)
 		}
+		pub(crate) fn do_get_exchange_rate(
+			pool_id: T::PoolId,
+			swap: bool,
+		) -> Result<Rate, DispatchError> {
+			let pair = Self::currency_pair(pool_id)?;
+			let pool_account = Self::account_id(&pool_id);
+			let pair = if swap { pair.swap() } else { pair };
+			let pool_base_asset_under_management =
+				T::Convert::convert(T::Assets::balance(pair.base, &pool_account));
+			let pool_quote_asset_under_management =
+				T::Convert::convert(T::Assets::balance(pair.quote, &pool_account));
+			Ok(Rate::checked_from_rational(
+				pool_base_asset_under_management,
+				pool_quote_asset_under_management,
+			)
+			.ok_or(ArithmeticError::Overflow)?)
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			let current_timestamp = T::Time::now();
+			Pools::<T>::iter()
```

Comment:
Do we run this logic every single block? That seems prohibitively expensive to me.

Reply:
We can avoid doing that if the interval is always the same for all pool. Which is currently the case so we can probably directly have the extrinsic to enable TWAP in pablo, and setup the initial price cumulative. This would allow the hook to execute that only once the TWAPInterval expired rather than iterating + checking

Reply:
We can avoid doing that if the interval is always the same for all pool. Which is currently the case so we can probably directly have the extrinsic to enable TWAP in pablo, and setup the initial price cumulative. This would allow the hook to execute that only once the TWAPInterval expired rather than iterating + checking

Reply:
@vivekvpandya the `EnabledTWAP` doesn't make sense anymore in Pablo as we have all pools under the same pallet. We can remove it and have a local map that tell whether a pool is enabled or not. With an extrinsic to enable + populate first price cumulative

Reply:
@vivekvpandya the `EnabledTWAP` doesn't make sense anymore in Pablo as we have all pools under the same pallet. We can remove it and have a local map that tell whether a pool is enabled or not. With an extrinsic to enable + populate first price cumulative

Reply:
:+1: